test/test_client.py

```diff
@@ -0,0 +1,77 @@
+from collections import Counter
+from unittest import TestCase
+import logging
+
+from mock import Mock, patch
+
+from apns2.client import APNsClient, CONCURRENT_STREAMS_SAFETY_MAXIMUM
+from apns2.payload import Payload
+
+class ClientTestCase(TestCase):
+    @classmethod
+    def setUpClass(cls):
+        logging.basicConfig()
+        cls.tokens = ["%064x" % i for i in range(10000)]
+        cls.notification = Payload(alert="Hello, world!")
+        cls.topic = "com.example.App"
+
+    
+    def setUp(self):
+        self.open_streams = 0
+        self.max_open_streams = 0
+        self.mock_results = None
+        
+        with patch("apns2.client.HTTP20Connection") as mock_connection_constructor, \
+            patch("apns2.client.init_context"):
+            self.mock_connection = Mock()
+            self.mock_connection.get_response.side_effect = self.mock_get_response
+            self.mock_connection.request.side_effect = self.mock_request
+            self.mock_connection.remote_settings.max_concurrent_streams = 500
+            mock_connection_constructor.return_value = self.mock_connection
+            self.client = APNsClient(cert_file=None)
+            
+    def mock_get_response(self, *dummy_args):
+        self.open_streams -= 1
+        if self.mock_results:
+            reason = self.mock_results[self.mock_connection.get_response.call_count - 1]
+            response = Mock(status=200 if reason == "Success" else 400)
+            response.read.return_value = '{"reason": "%s"}' % reason
+            return response
+        else:
+            return Mock(status=200)
```

Comment:
`mock_results` is way for a test to say "get_response mock, please return the following values [different value every time you're called". In all other cases, when `mock_results` is `None` (which is the case in all tests but `test_send_notification_batch_reports_different_results`), get_response always returns success.